### PR TITLE
Add sndio mixer, fix building on BSD

### PR DESCRIPTION
**Pull request by madroach**
Building on any unix with ``sin_size`` in ``sockaddr_in`` is broken.
I fix this not by using additional ``#ifdef``s, but by initializing the anonymous ``sockaddr_in`` with explicit field names. The ``sin_size`` field does not require initialization and is implicitly initialized to ``0``.

### CODE REVIEW

File: src/net/IPv4Address.hxx [hunk 1]
```diff
@@ -81,13 +81,9 @@ class IPv4Address {
...
 						      uint16_t port) {
 		return {
-#if defined(__APPLE__)
-			sizeof(struct sockaddr_in),
-#endif
-			AF_INET,
-			ToBE16(port),
-			address,
-			{},
+			.sin_family = AF_INET,
```
Comment (MaxKellermann) [inline src/net/IPv4Address.hxx]: This is not portable. It is illegal C++ syntax, and works only because it uses a GCC extension.

File: src/output/plugins/SndioOutputPlugin.cxx [hunk 2]
```diff
@@ -44,30 +45,21 @@ static constexpr unsigned MPD_SNDIO_BUFFER_TIME_MS = 250;
 
 static constexpr Domain sndio_output_domain("sndio_output");
 
-class SndioOutput final : AudioOutput {
```
Comment (MaxKellermann) [inline src/output/plugins/SndioOutputPlugin.cxx]: This removal is obscure and hard to review, because it moves code and modifies it at the same time. Please make the move a separate commit.

File: src/output/plugins/SndioOutputPlugin.cxx [hunk 3]
```diff
@@ -163,9 +158,34 @@ SndioOutput::Play(const void *chunk, size_t size)
...
+void
+SndioOutput::RegisterMixerListener(Mixer *_mixer, MixerListener *_listener) {
+	mixer = _mixer;
+	listener = _listener;
+}
+
 const struct AudioOutputPlugin sndio_output_plugin = {
 	"sndio",
 	sndio_test_default_device,
-	&SndioOutput::Create,
```
Comment (MaxKellermann) [inline src/output/plugins/SndioOutputPlugin.cxx]: Why this change?

File: src/output/plugins/SndioOutputPlugin.cxx [hunk 4]
```diff
@@ -163,9 +158,34 @@ SndioOutput::Play(const void *chunk, size_t size)
...
 
+void
+SndioOutput::SetVolume(unsigned int volume) {
+	sio_setvol(sio_hdl, volume * SIO_MAXVOL / 100);
+}
+
+void
+SndioOutput::VolumeChanged(unsigned int _raw_volume) {
+	raw_volume = _raw_volume;
+	if (listener != NULL && mixer != NULL) {
```
Comment (MaxKellermann) [inline src/output/plugins/SndioOutputPlugin.cxx]: Use `nullptr` instead of the deprecated `NULL`.

File: src/output/plugins/SndioOutputPlugin.cxx [hunk 5]
```diff
@@ -140,6 +132,9 @@ SndioOutput::Open(AudioFormat &audio_format)
 		throw std::runtime_error("Requested audio params cannot be satisfied");
 	}
 
+	sio_setvol(sio_hdl, raw_volume);
```
Comment (MaxKellermann) [inline src/output/plugins/SndioOutputPlugin.cxx]: Why modify the volume while opening the output? No other plugin does that.

File: src/output/plugins/SndioOutputPlugin.cxx [hunk 6]
```diff
@@ -44,30 +45,21 @@ static constexpr unsigned MPD_SNDIO_BUFFER_TIME_MS = 250;
...
-	struct sio_hdl *sio_hdl;
-
-public:
-	SndioOutput(const ConfigBlock &block);
-
-	static AudioOutput *Create(EventLoop &,
-				   const ConfigBlock &block) {
-		return new SndioOutput(block);
-	}
+void SndioMixerCallback(void *arg, unsigned int volume) {
```
Comment (MaxKellermann) [inline src/output/plugins/SndioOutputPlugin.cxx]: Should be static.

File: src/output/plugins/SndioOutputPlugin.cxx [hunk 7]
```diff
@@ -44,30 +45,21 @@ static constexpr unsigned MPD_SNDIO_BUFFER_TIME_MS = 250;
...
+	return new SndioOutput(block);
+}
 
 SndioOutput::SndioOutput(const ConfigBlock &block)
 	:AudioOutput(0),
 	 device(block.GetBlockValue("device", SIO_DEVANY)),
 	 buffer_time(block.GetBlockValue("buffer_time",
-					 MPD_SNDIO_BUFFER_TIME_MS))
+					 MPD_SNDIO_BUFFER_TIME_MS)),
+	 raw_volume(SIO_MAXVOL)
```
Comment (MaxKellermann) [inline src/output/plugins/SndioOutputPlugin.cxx]: Use C++11 initializer.

File: src/output/plugins/SndioOutputPlugin.cxx [hunk 8]
```diff
@@ -163,9 +158,34 @@ SndioOutput::Play(const void *chunk, size_t size)
...
+	raw_volume = _raw_volume;
+	if (listener != NULL && mixer != NULL) {
+		listener->OnMixerVolumeChanged(*mixer,
+		    _raw_volume * 100 / SIO_MAXVOL);
+	}
+}
+
+unsigned int
+SndioOutput::GetVolume() {
+	return (raw_volume * 100 / SIO_MAXVOL);
```
Comment (MaxKellermann) [inline src/output/plugins/SndioOutputPlugin.cxx]: Duplicate formula; this code is already present in `VolumeChanged()`. Maybe better remember the MPD-style value instead of the raw value?

File: src/output/plugins/SndioOutputPlugin.cxx (hunk 6 above)
Comment (MaxKellermann) [inline src/output/plugins/SndioOutputPlugin.cxx]: Coding style violation.

File: src/output/plugins/SndioOutputPlugin.cxx [hunk 9]
```diff
@@ -163,9 +158,34 @@ SndioOutput::Play(const void *chunk, size_t size)
...
+SndioOutput::VolumeChanged(unsigned int _raw_volume) {
+	raw_volume = _raw_volume;
+	if (listener != NULL && mixer != NULL) {
+		listener->OnMixerVolumeChanged(*mixer,
+		    _raw_volume * 100 / SIO_MAXVOL);
+	}
+}
+
+unsigned int
+SndioOutput::GetVolume() {
```
Comment (MaxKellermann) [inline src/output/plugins/SndioOutputPlugin.cxx]: Another coding style violation.

File: src/output/plugins/SndioOutputPlugin.hxx [hunk 10]
```diff
@@ -20,6 +20,38 @@
...
 
+#include "../OutputAPI.hxx"
+#include "mixer/Listener.hxx"
+
 extern const struct AudioOutputPlugin sndio_output_plugin;
 
+class PulseMixer;
+
+class SndioOutput final : AudioOutput {
+	friend class SndioMixer;
```
Comment (MaxKellermann) [inline src/output/plugins/SndioOutputPlugin.hxx]: No `friend`, please, unless ultimately necessary! `friend` can access any private, and is obscure. Provide an API instead of friending other classes.

File: src/output/plugins/SndioOutputPlugin.hxx [hunk 11]
```diff
@@ -20,6 +20,38 @@
...
+	friend void SndioMixerCallback(void *arg, unsigned int volume);
+
+	Mixer *mixer = NULL;
+	MixerListener *listener = NULL;
```
Comment (MaxKellermann) [inline src/output/plugins/SndioOutputPlugin.hxx]: `nullptr` ...

File: src/net/IPv4Address.hxx (hunk 1 above)
Comment (madroach) [reply src/net/IPv4Address.hxx]: Oh, I'm sorry. I don't know much C++.
I didn't notice this feature was not yet standardized.

Then how would you fix this? Add a feature test macro?
I'd suggest removing the constexpr optimization altogether. This is no critical code path.

File: src/output/plugins/SndioOutputPlugin.cxx (hunk 5 above)
Comment (madroach) [reply src/output/plugins/SndioOutputPlugin.cxx]: Because continuing playback after stop will open a new sndiod stream which will have a new 100% volume assigned to it. Could be a loud surprise.

File: src/output/plugins/SndioOutputPlugin.cxx (hunk 8 above)
Comment (madroach) [reply src/output/plugins/SndioOutputPlugin.cxx]: I thought about this, but honestly don't remember the reason right now. I believe the code would be duplicated elsewhere when storing the mpd value, or rounding errors would appear when a third party changes the volume.

**Comment by madroach** [comment]
Thanks for your review and feedback. Maybe I'll find some time to rework the diff on the weekend.

File: src/output/plugins/SndioOutputPlugin.cxx (hunk 5 above)
Comment (MaxKellermann) [reply src/output/plugins/SndioOutputPlugin.cxx]: So there is nothing inside sndio which remembers the volume? Then this line is ok, but add a code comment explaining it.

File: src/output/plugins/SndioOutputPlugin.cxx (hunk 8 above)
Comment (MaxKellermann) [reply src/output/plugins/SndioOutputPlugin.cxx]: OK, move the formula to a small function. Don't duplicate code. This is maintenance hell.

**Comment by madroach** [comment]
I reworked the diff to address the requested changes.

**Comment by MaxKellermann** [comment]
/home/max/git/mpd/src/mixer/plugins/SndioMixerPlugin.cxx: In constructor 'SndioMixer::SndioMixer(SndioOutput&, MixerListener&)':
/home/max/git/mpd/src/mixer/plugins/SndioMixerPlugin.cxx:31:3: error: declaration of 'listener' shadows a member of 'SndioMixer' [-Werror=shadow]
   :Mixer(sndio_mixer_plugin, listener), output(output)
   ^
In file included from /home/max/git/mpd/src/mixer/plugins/SndioMixerPlugin.cxx:20:0:
/home/max/git/mpd/src/mixer/MixerInternal.hxx:34:17: note: shadowed declaration is here
  MixerListener &listener;
                 ^~~~~~~~
/home/max/git/mpd/src/mixer/plugins/SndioMixerPlugin.cxx:31:3: error: declaration of 'output' shadows a member of 'SndioMixer' [-Werror=shadow]
   :Mixer(sndio_mixer_plugin, listener), output(output)
   ^
/home/max/git/mpd/src/mixer/plugins/SndioMixerPlugin.cxx:27:15: note: shadowed declaration is here
  SndioOutput &output;
               ^~~~~~

**Comment by MaxKellermann** [comment]
I've merged the first two patches, thanks for that so far.

**Comment by madroach** [comment]
renamed the shadowing parameter to ``_listener``.
clang doesn't know ``-Wshadow``. That's why I didn't get this error.

**Comment by MaxKellermann** [comment]
The other shadow remains:

/home/max/git/mpd/src/mixer/plugins/SndioMixerPlugin.cxx: In constructor 'SndioMixer::SndioMixer(SndioOutput&, MixerListener&)':
/home/max/git/mpd/src/mixer/plugins/SndioMixerPlugin.cxx:31:3: error: declaration of 'output' shadows a member of 'SndioMixer' [-Werror=shadow]
   :Mixer(sndio_mixer_plugin, _listener), output(output)
   ^
/home/max/git/mpd/src/mixer/plugins/SndioMixerPlugin.cxx:27:15: note: shadowed declaration is here

**Comment by madroach** [comment]
Thanks for your review and merging!
